index.js

```diff
@@ -31,24 +33,98 @@ exports.register = (server, options) => {
   // expose sentry client at server.plugins['hapi-sentry'].client
   server.expose('client', Sentry);
 
-  server.ext({
-    type: 'onRequest',
-    method(request, h) {
-      if (opts.useDomainPerRequest) {
-        // Sentry looks for current hub in active domain
-        // Therefore simply by creating&entering domain Sentry will create
-        // request scoped hub for breadcrumps and other scope metadata
-        request.__sentryDomain = domain.create();
-        request.__sentryDomain.enter();
-      }
+  // Set up request interceptor for creating and destroying a domain on each request
+  // It'll wrap the request handler returned by the _dispatch function in HAPI core
+  // allowing it to create the domain before any of HAPIs processing starts
+  // thus allowing us to use the normal HAPI extensions to add context to Sentry scopes
+  // https://github.com/hapijs/hapi/blob/c95985e225fa09c4b640a887ccb4be46dbe265bc/lib/core.js#L507-L538
+  function interceptor(next, req, res, ...args) {
+    const local = domain.create(); // Create domain to hold context for request
+    local.add(req);
+    local.add(res);
+
+    let rtn;
+
+    local.run(() => { // Create new scope for request
+      const currentHub = Sentry.getCurrentHub();
+
+      currentHub.configureScope(scope => {
+        scope.addEventProcessor(_sentryEvent => {
+          // format a sentry event from the request and triggered event
+          const sentryEvent = Sentry.Handlers.parseRequest(_sentryEvent, req);
+
+          // overwrite events request url if a baseUrl is provided
+          if (opts.baseUri) {
+            if (opts.baseUri.slice(-1) === '/') opts.baseUri = opts.baseUri.slice(0, -1);
+            sentryEvent.request.url = opts.baseUri + req.path;
+          }
+
+          // some SDK identifier
+          sentryEvent.sdk = { name: 'sentry.javascript.node.hapi', version };
+          return sentryEvent;
+        });
+      });
 
-      // attach a new scope to each request for breadcrumbs/tags/extras/etc capturing
-      request.sentryScope = new Sentry.Scope();
+      rtn = next.apply(this, [null, req, res].concat(args));
+    });
 
-      return h.continue;
-    },
+    return rtn;
+  }
+
+  // Wrap HAPI core _dispatch function. This function is primary entry point into HAPI for an
+  // external request. It's a factory that returns Node request handlers
+  // https://github.com/hapijs/hapi/blob/c95985e225fa09c4b640a887ccb4be46dbe265bc/lib/core.js#L505-L539
+  if (server._core && server._core._dispatch && !server._core._dispatch.__wrapped) {
```

Comment:
what happens if this is not the case? the plugin does not work right?

If that is the case, can do an early return here? either throw an error or log that the plugin now does nothing (error is better i think)

Reply:
The plugin will still work. But `server.inject()` calls won't get request context separation. Normal request arriving from outside Node via `http.Server` will still work.

Reply:
Thanks for explaining.

This comment made me think that everything will break if that if does not evaluate to true:
> This function is primary entry point into HAPI for an external request

Can we add that as a comment? Like "This is only needed for request context separation when using server.inject()" or something like that.

Reply:
Yeah I've just re-written this comment include pretty much what you've suggested.

Reply:
yea looks good